File: 백준/Gold/15683. 감시/감시.cc
    #include <iostream>
#include <vector>

using namespace std;

/**
접근
- 사각지대의 최소갯수구하기
-CCTV마다 방향이 있고, 벽을 통과할 수 없음.

1. cctv 방향 경우의 수
4(방향)^8(최대CCTV) = 2^16 = 2^10 * 2^6 = 1024 * 64 = 약 60000

2. board업데이트
cctv마다 벽을 만날때까지 WATCHED로 변경
(CCTV는 건너뜀, 맵을 벗어나면 끝냄)
4(방향) * 8(최대맵) * 8(최대CCTV) = 256

3. 사각지대갯수 카운트
아직 0인것 세기
8*8

총 시간복잡도 60000 * (256 + 64) 정도로 가능
*/

int n, m;
const int MX = 10;
int board[MX][MX];
int board2[MX][MX];

const int WALL = 6;
const int WATCHED = 7;

int tmp[MX];
vector<pair<int,int>> cctvs;
vector<string> cases;


int dx[4] = {1, 0, -1, 0};
int dy[4] = {0, -1, 0, 1};

void update(int x, int y, int dir){
    dir %= 4;
    int nx = x;
    int ny = y;
    while(true){
        nx += dx[dir];
        ny += dy[dir];
        if(nx < 0 || nx >= n || ny < 0 || ny >= m) break;
        if(board2[nx][ny] == WALL) break;
        if(board2[nx][ny] != 0) continue;
        board2[nx][ny] = WATCHED;
    }
}

void monitor(int x, int y, int dir){
    int cctv = board2[x][y];
  
    if(cctv == 1){
        update(x, y, dir);
    } 
    else if(cctv == 2){
        update(x, y, dir);
        update(x, y, dir+2);
    }
    else if(cctv == 3){
        update(x, y, dir);
        update(x, y, dir+3);
    }
    else if(cctv == 4){
        update(x, y, dir);
        update(x, y, dir+2);
        update(x, y, dir+3);
    }
    else if(cctv == 5){
        update(x, y, dir);
        update(x, y, dir+1);
        update(x, y, dir+2);
        update(x, y, dir+3);
    }
}

void recur(int cur){
    if(cur == cctvs.size()){
        string str = "";
        for(int i = 0; i < cctvs.size(); i++) {
            str += to_string(tmp[i]);
        }
        cases.push_back(str);
        return;
    }
    for(int i = 0; i < 4; i++){
        tmp[cur] = i;
        recur(cur+1);
    }
}

int main()
{
    cin >> n >> m;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < m; j++){
            cin >> board[i][j];
            if(0 < board[i][j] && board[i][j] < 6) {
                cctvs.push_back({i,j});
            }
        }
    }
    recur(0);
    int mn = 1e9;
    // 1. cctv방향 경우의 수에 대해
    for(string cc : cases){
        int cnt = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                board2[i][j] = board[i][j];
            }
        }
        
        // 2. cctv를 돌면서 board update
        for(int i = 0; i < cctvs.size(); i++){
            int x = cctvs[i].first;
            int y = cctvs[i].second;
            int dir = (int) (cc[i] - '0');
          
            monitor(x, y, dir);
        }
  
        // 3. 사각지대 카운트
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(board2[i][j] == 0) cnt++;
            }
        }
        
        mn = min(cnt, mn);
    }

    cout << mn;

    return 0;
}

File: 백준/Gold/15683. 감시/감시.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
	static int R, C;
	static int ori[][] = new int[10][10];
	static int map[][] = new int[10][10];
	// 오, 아, 왼, 위
	static int dr[] = new int[] { 0, 1, 0, -1 };
	static int dc[] = new int[] { 1, 0, -1, 0 };
	static int rotate[] = new int[10];
	static int mn = Integer.MAX_VALUE;
	static ArrayList<Integer[]> cctvs = new ArrayList<>();

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		R = Integer.parseInt(st.nextToken());
		C = Integer.parseInt(st.nextToken());

		for (int r = 0; r < R; r++) {
			st = new StringTokenizer(br.readLine());
			for (int c = 0; c < C; c++) {
				ori[r][c] = Integer.parseInt(st.nextToken());
				int type = ori[r][c];
				if (ori[r][c] >= 1 && ori[r][c] <= 5) {
					cctvs.add(new Integer[] { r, c, type });
				}
			}
		}

		recur(0);

		System.out.println(mn);
	}

	static void recur(int cur) {
		if (cur == cctvs.size()) {
			map = copy(ori);
			for (int i = 0; i < cur; i++) {
				Integer[] cctv = cctvs.get(i);
				scan(cctv, rotate[i]);
			}

			int cnt = 0;
			for (int r = 0; r < R; r++) {
				for (int c = 0; c < C; c++) {
					if (map[r][c] == 0)
						cnt++;
				}
			}

			mn = Math.min(cnt, mn);
			return;
		}
		for (int i = 0; i < 4; i++) {
			rotate[cur] = i;
			recur(cur + 1);
		}
	}

	static void scan(Integer[] cctv, int rotate) {
		int r = cctv[0];
		int c = cctv[1];
		int type = cctv[2];

		// 오, 아, 왼, 위
		int right = 0;
		int down = 1;
		int left = 2;
		int up = 3;

		if (type == 1) {
			scan(r, c, (right + rotate) % 4);
		} else if (type == 2) {
			scan(r, c, (right + rotate) % 4);
			scan(r, c, (left + rotate) % 4);
		} else if (type == 3) {
			scan(r, c, (up + rotate) % 4);
			scan(r, c, (right + rotate) % 4);
		} else if (type == 4) {
			scan(r, c, (up + rotate) % 4);
			scan(r, c, (right + rotate) % 4);
			scan(r, c, (left + rotate) % 4);
		} else if (type == 5) {
			scan(r, c, (up + rotate) % 4);
			scan(r, c, (right + rotate) % 4);
			scan(r, c, (left + rotate) % 4);
			scan(r, c, (down + rotate) % 4);
		}
	}

	static void scan(int r, int c, int dir) {
		int nr = r;
		int nc = c;

		while (true) {
			nr += dr[dir];
			nc += dc[dir];
			if (nr < 0 || nr >= R || nc < 0 || nc >= C) {
				break;
			}
			if (map[nr][nc] == 6) { // wall
				break;
			}
			if (map[nr][nc] == 0) {
				map[nr][nc] = 9;
			}
		}
	}

	static int[][] copy(int[][] map) {
		int tmp[][] = new int[R][C];

		for (int i = 0; i < R; i++) {
			System.arraycopy(map[i], 0, tmp[i], 0, C);
		}

		return tmp;
	}
}